src/modules/ekf2/EKF/terrain_estimator.cpp

```diff
@@ -43,94 +43,159 @@
 
-	return initialized;
+	updateTerrainValidity();
 }
 
-void Ekf::runTerrainEstimator()
+void Ekf::predictHagl()
 {
-	// If we are on ground, store the local position and time to use as a reference
-	if (!_control_status.flags.in_air) {
-		_last_on_ground_posD = _state.pos(2);
+	// predict the state variance growth where the state is the vertical position of the terrain underneath the vehicle
+
+	// process noise due to errors in vehicle height estimate
+	_terrain_var += sq(_imu_sample_delayed.delta_vel_dt * _params.terrain_p_noise);
+
+	// process noise due to terrain gradient
+	_terrain_var += sq(_imu_sample_delayed.delta_vel_dt * _params.terrain_gradient)
+			* (sq(_state.vel(0)) + sq(_state.vel(1)));
+
+	// limit the variance to prevent it becoming badly conditioned
+	_terrain_var = math::constrain(_terrain_var, 0.0f, 1e4f);
+}
+
+void Ekf::controlHaglRngFusion()
+{
+	if (!(_params.terrain_fusion_mode & TerrainFusionMask::TerrainFuseRangeFinder)
+	    || _control_status.flags.rng_fault) {
+
+		stopHaglRngFusion();
+		return;
 	}
 
-	// Perform initialisation check and
-	// on ground, continuously reset the terrain estimator
-	if (!_terrain_initialised || !_control_status.flags.in_air) {
-		_terrain_initialised = initHagl();
+	if (_range_sensor.isDataHealthy()) {
+		const bool continuing_conditions_passing = _control_status.flags.in_air;
+		//const bool continuing_conditions_passing = _control_status.flags.in_air && !_control_status.flags.rng_hgt; // TODO: should not be fused when using range height
+		const bool starting_conditions_passing = continuing_conditions_passing && _range_sensor.isRegularlySendingData();
 
-	} else {
+		_time_last_healthy_rng_data = _time_last_imu;
 
-		// predict the state variance growth where the state is the vertical position of the terrain underneath the vehicle
+		if (_hagl_sensor_status.flags.range_finder) {
+			if (continuing_conditions_passing) {
+				fuseHaglRng();
 
-		// process noise due to errors in vehicle height estimate
-		_terrain_var += sq(_imu_sample_delayed.delta_vel_dt * _params.terrain_p_noise);
+				// We have been rejecting range data for too long
+				const uint64_t timeout = static_cast<uint64_t>(_params.terrain_timeout * 1e6f);
+				const bool is_fusion_failing = isTimedOut(_time_last_hagl_fuse, timeout);
 
-		// process noise due to terrain gradient
-		_terrain_var += sq(_imu_sample_delayed.delta_vel_dt * _params.terrain_gradient)
-				* (sq(_state.vel(0)) + sq(_state.vel(1)));
+				if (is_fusion_failing) {
+					if (_range_sensor.getDistBottom() > 2.f * _params.rng_gnd_clearance) {
+						// Data seems good, attempt a reset
+						resetHaglRng();
 
-		// limit the variance to prevent it becoming badly conditioned
-		_terrain_var = math::constrain(_terrain_var, 0.0f, 1e4f);
+					} else if (starting_conditions_passing) {
+						// The sensor can probably not detect the ground properly
+						// declare the sensor faulty and stop the fusion
+						_control_status.flags.rng_fault = true;
```

Comment:
Should we be resetting these flags when landed (and disarmed)? 

I'm worried about potential edge cases where there are differences between back to back flights vs full reset before flying again. I don't want to people to develop superstitions or the habit of "turning if off and on again". 

Alternatively we could think about higher level infrastructure where a fault like this is flagged (and stored) and the system would then require you to do something to explicitly address it before it can be used again. For example if `_control_status.flags.mag_fault` is set we could set something that actually requires the user to perform a calibration before it's re-enabled.

Reply:
Ok, done in [18cd3ce](https://github.com/PX4/PX4-Autopilot/pull/19148/commits/18cd3ce4e0257512a2b1ffb5083d7e60c927d07e)

Reply:
I should also add user reporting and a check in commander like for the `mag_fault`; but let's keep that PR simple. I can also remove the detection feature from that PR and do it separately if needed